File: internal/store/memory/chat-sessions.go
package memory

import (
	"context"
	"slices"
	"time"

	"github.com/oklog/ulid/v2"
	"github.com/rakunlabs/at/internal/service"
	"github.com/rakunlabs/query"
)

// ─── Chat Session CRUD ───

func (m *Memory) ListChatSessions(_ context.Context, q *query.Query) (*service.ListResult[service.ChatSession], error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	result := make([]service.ChatSession, 0, len(m.chatSessions))
	for _, s := range m.chatSessions {
		result = append(result, s)
	}

	slices.SortFunc(result, func(a, b service.ChatSession) int {
		if a.CreatedAt < b.CreatedAt {
			return -1
		}
		if a.CreatedAt > b.CreatedAt {
			return 1
		}
		return 0
	})

	return paginate(result, q), nil
}

func (m *Memory) GetChatSession(_ context.Context, id string) (*service.ChatSession, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	s, ok := m.chatSessions[id]
	if !ok {
		return nil, nil
	}

	return &s, nil
}

func (m *Memory) GetChatSessionByTaskID(_ context.Context, taskID string) (*service.ChatSession, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	for _, s := range m.chatSessions {
		if s.TaskID == taskID && s.TaskID != "" {
			return &s, nil
		}
	}

	return nil, nil
}

func (m *Memory) GetChatSessionByPlatform(_ context.Context, platform, platformUserID, platformChannelID string) (*service.ChatSession, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	for _, s := range m.chatSessions {
		if s.Config.Platform == platform && s.Config.PlatformUserID == platformUserID && s.Config.PlatformChannelID == platformChannelID {
			return &s, nil
		}
	}

	return nil, nil
}

func (m *Memory) CreateChatSession(_ context.Context, session service.ChatSession) (*service.ChatSession, error) {
	id := ulid.Make().String()
	now := time.Now().UTC().Format(time.RFC3339)

	rec := service.ChatSession{
		ID:             id,
		AgentID:        session.AgentID,
		TaskID:         session.TaskID,
		OrganizationID: session.OrganizationID,
		Name:           session.Name,
		Config:         session.Config,
		CreatedAt:      now,
		UpdatedAt:      now,
		CreatedBy:      session.CreatedBy,
		UpdatedBy:      session.UpdatedBy,
	}

	m.mu.Lock()
	m.chatSessions[id] = rec
	m.mu.Unlock()

	return &rec, nil
}

func (m *Memory) UpdateChatSession(_ context.Context, id string, session service.ChatSession) (*service.ChatSession, error) {
	now := time.Now().UTC().Format(time.RFC3339)

	m.mu.Lock()
	defer m.mu.Unlock()

	existing, ok := m.chatSessions[id]
	if !ok {
		return nil, nil
	}

	existing.Name = session.Name
	existing.Config = session.Config
	existing.UpdatedAt = now
	existing.UpdatedBy = session.UpdatedBy
	if session.AgentID != "" {
		existing.AgentID = session.AgentID
	}
	if session.TaskID != "" {
		existing.TaskID = session.TaskID
	}
	if session.OrganizationID != "" {
		existing.OrganizationID = session.OrganizationID
	}

	m.chatSessions[id] = existing

	return &existing, nil
}

func (m *Memory) DeleteChatSession(_ context.Context, id string) error {
	m.mu.Lock()
	delete(m.chatSessions, id)
	// Also delete messages for this session.
	delete(m.chatMessages, id)
	m.mu.Unlock()

	return nil
}

func (m *Memory) ListChatMessages(_ context.Context, sessionID string) ([]service.ChatMessage, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()

	msgs := m.chatMessages[sessionID]
	result := make([]service.ChatMessage, len(msgs))
	copy(result, msgs)

	return result, nil
}

func (m *Memory) CreateChatMessage(_ context.Context, msg service.ChatMessage) (*service.ChatMessage, error) {
	id := ulid.Make().String()
	now := time.Now().UTC().Format(time.RFC3339)

	rec := service.ChatMessage{
		ID:        id,
		SessionID: msg.SessionID,
		Role:      msg.Role,
		Data:      msg.Data,
		CreatedAt: now,
	}

	m.mu.Lock()
	m.chatMessages[msg.SessionID] = append(m.chatMessages[msg.SessionID], rec)
	m.mu.Unlock()

	return &rec, nil
}

func (m *Memory) CreateChatMessages(_ context.Context, msgs []service.ChatMessage) error {
	now := time.Now().UTC().Format(time.RFC3339)

	m.mu.Lock()
	defer m.mu.Unlock()

	for _, msg := range msgs {
		id := ulid.Make().String()
		rec := service.ChatMessage{
			ID:        id,
			SessionID: msg.SessionID,
			Role:      msg.Role,
			Data:      msg.Data,
			CreatedAt: now,
		}
		m.chatMessages[msg.SessionID] = append(m.chatMessages[msg.SessionID], rec)
	}

	return nil
}

func (m *Memory) DeleteChatMessages(_ context.Context, sessionID string) error {
	m.mu.Lock()
	delete(m.chatMessages, sessionID)
	m.mu.Unlock()
	return nil
}

File: _ui/src/lib/api/chat-sessions.ts
import axios from 'axios';
import { authFetch as fetch } from './transport';
import type { ListResult, ListParams } from './types';

const api = axios.create({ baseURL: 'api/v1' });

export interface ChatSessionConfig {
  platform?: string;
  platform_user_id?: string;
  platform_channel_id?: string;
  bot_config_id?: string;
  active_task_id?: string;
  history_limit?: number;
  task_discussion_mode?: boolean;
  disable_task_result_sync?: boolean;
}

export interface ListChatSessionsParams extends ListParams {
  /** Filter sessions to only those tied to the given BotConfig. */
  bot_config_id?: string;
  /** Filter sessions to only those for a given platform (telegram, discord, …). */
  platform?: string;
}

export interface ChatSession {
  id: string;
  agent_id: string;
  task_id?: string;
  organization_id?: string;
  name: string;
  config: ChatSessionConfig;
  created_at: string;
  updated_at: string;
  created_by: string;
  updated_by: string;
}

export interface ChatMessageData {
  content: any;
  tool_calls?: any;
  tool_call_id?: string;
  attachments?: ChatAttachment[];
}

export interface ChatAttachment {
  name: string;
  media_type: string;
  data: string;
}

export interface ChatMessage {
  id: string;
  session_id: string;
  role: string;
  data: ChatMessageData;
  created_at: string;
}

export async function listChatSessions(params?: ListChatSessionsParams): Promise<ListResult<ChatSession>> {
  const res = await api.get<ListResult<ChatSession>>('/chat/sessions', { params });
  return res.data;
}

export async function getChatSession(id: string): Promise<ChatSession> {
  const res = await api.get<ChatSession>(`/chat/sessions/${id}`);
  return res.data;
}

export async function createChatSession(data: Partial<ChatSession>): Promise<ChatSession> {
  const res = await api.post<ChatSession>('/chat/sessions', data);
  return res.data;
}

export async function updateChatSession(id: string, data: Partial<ChatSession>): Promise<ChatSession> {
  const res = await api.put<ChatSession>(`/chat/sessions/${id}`, data);
  return res.data;
}

export async function deleteChatSession(id: string): Promise<void> {
  await api.delete(`/chat/sessions/${id}`);
}

export async function clearChatMessages(sessionId: string): Promise<void> {
  await api.delete(`/chat/sessions/${sessionId}/messages`);
}

export interface ListChatMessagesOptions {
  /** Return only the most recent N messages (chronological order). */
  limit?: number;
  /** Page messages strictly older than this message ID (scroll-up lazy load). */
  beforeId?: string;
}

export async function listChatMessages(
  sessionId: string,
  opts?: ListChatMessagesOptions
): Promise<ChatMessage[]> {
  const params: Record<string, string | number> = {};
  if (opts?.limit) params.limit = opts.limit;
  if (opts?.beforeId) params.before_id = opts.beforeId;
  const res = await api.get<ChatMessage[]>(`/chat/sessions/${sessionId}/messages`, { params });
  return res.data;
}

/** Send a message to a chat session and receive SSE events. Returns an AbortController. */
export function sendMessage(
  sessionId: string,
  content: string,
  onEvent: (event: any) => void,
  onError: (error: string) => void,
  onDone: () => void | Promise<void>,
  attachments: ChatAttachment[] = [],
): AbortController {
  const controller = new AbortController();

  const basePath = document.querySelector('base')?.getAttribute('href') || '';

  fetch(`${basePath}api/v1/chat/sessions/${sessionId}/messages`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ content, attachments }),
    signal: controller.signal,
  })
    .then(async (response) => {
      if (!response.ok) {
        const text = await response.text();
        let message = text;
        try { message = JSON.parse(text)?.message || text; } catch { /* Non-JSON proxy response. */ }
        onError(message || `HTTP ${response.status}`);
        return;
      }

      await consumeChatEvents(response, onEvent, onDone);
    })
    .catch((err) => {
      if (err.name !== 'AbortError') {
        onError(err.message || 'Network error');
      }
    });

  return controller;
}

/** Decode complete SSE frames, including split UTF-8 and a final unterminated frame.
 * A disconnected stream must never look like a completed (persisted) answer. */
export async function consumeChatEvents(
  response: Response,
  onEvent: (event: any) => void,
  onDone: () => void | Promise<void>,
): Promise<void> {
  const reader = response.body?.getReader();
  if (!reader) throw new Error('No response body');
  const decoder = new TextDecoder();
  let buffer = '';
  let data: string[] = [];
  let eventName = '';
  const dispatch = async () => {
    if (!data.length) return false;
    const payload = data.join('\n');
    data = [];
    const event = JSON.parse(payload);
    if (event.error || event.type === 'error' || eventName === 'error') {
      throw new Error(typeof event.error === 'string' ? event.error : event.error?.message || event.message || 'The agent could not complete this response');
    }
    eventName = '';
    if (event.type === 'done') { await onDone(); return true; }
    onEvent(event);
    return false;
  };
  const line = async (value: string) => {
    if (!value) {
      const complete = await dispatch();
      eventName = '';
      return complete;
    }
    if (value.startsWith('data:')) data.push(value.slice(5).replace(/^ /, ''));
    if (value.startsWith('event:')) eventName = value.slice(6).trim();
    return false;
  };
  try {
    while (true) {
      const { done, value } = await reader.read();
      buffer += done ? decoder.decode() : decoder.decode(value, { stream: true });
      let end: number;
      while ((end = buffer.indexOf('\n')) !== -1) {
        const current = buffer.slice(0, end).replace(/\r$/, '');
        buffer = buffer.slice(end + 1);
        if (await line(current)) return;
      }
      if (done) {
        if (buffer && await line(buffer.replace(/\r$/, ''))) return;
        if (await dispatch()) return;
        throw new Error('Connection ended before the response completed. Received text is preserved; retry when ready.');
      }
    }
  } finally {
    await reader.cancel().catch(() => {});
    reader.releaseLock();
  }
}

/** Send a tool confirmation (approve or reject) for a pending tool call. */
export async function confirmToolCall(
  sessionId: string,
  toolId: string,
  approved: boolean,
): Promise<void> {
  await api.post(`/chat/sessions/${sessionId}/confirm`, {
    tool_id: toolId,
    approved,
  });
}
